zaza/openstack/charm_tests/keystone/tests.py

```diff
@@ -436,6 +444,62 @@ def _find_keystone_v3_user(self, username, domain):
         )
         return None
 
+    def _find_keystone_v3_group(self, group, domain):
+        """Find a group within a specified keystone v3 domain.
+
+        :param str group: Group to search for in keystone
+        :param str domain: username selected from which domain
+        :return: return group if found
+        :rtype: Optional[str]
+        """
+        for ip in self.keystone_ips:
+            logging.info('Keystone IP {}'.format(ip))
+            session = openstack_utils.get_keystone_session(
+                openstack_utils.get_overcloud_auth(address=ip))
+            client = openstack_utils.get_keystone_session_client(session)
+
+            domain_groups = client.groups.list(
+                domain=client.domains.find(name=domain).id
+            )
+
+            for searched_group in domain_groups:
+                if searched_group.name.lower() == group.lower():
+                    return searched_group
+
+        logging.debug(
+            "Group {} was not found. Returning None.".format(group)
+        )
+        return None
+
+    def _find_keystone_v3_user_in_group(self, user, group, domain):
+        """Find a user in group group within a specified keystone v3 domain.
+
+        :param str user: User to search for in keystone
+        :param str group: Group to search for in keystone
+        :param str domain: username/group selected from which domain
+        :return: return group if found
+        :rtype: Optional[str]
+        """
+        for ip in self.keystone_ips:
+            logging.info('Keystone IP {}'.format(ip))
+            session = openstack_utils.get_keystone_session(
+                openstack_utils.get_overcloud_auth(address=ip))
+            client = openstack_utils.get_keystone_session_client(session)
+
+            domain_users = client.users.list(
+                domain=client.domains.find(name=domain).id,
+                group=self._find_keystone_v3_group(group, domain).id,
```

Comment:
Apart from this line it's the same implementation as `_find_keystone_v3_user()`. Couldn't we have extended that function with an optional `group` argument instead?

Reply:
hmm, nice catch, thanks, I will sort it, and push a new update later.

I did think at one point, that there was some repetitiveness, but couldn't fathom it :)